### DISCUSSION TITLE
Replace function or method call in wrapped `if` statement

### DISCUSSION THREAD

**Discussion by mglaman** [rectorphp/rector#7860] [Q&A] [answered]
I need a Rector rule which takes a function or method call and wraps it in a conditional `if` and `else`.

When testing, I tried just returning an `if` and got the following error:

```
LogicException: Trying to replace expression (Expr_FuncCall) with statement (Stmt_If)
```

Are the examples in Rector for replacing expressions and wrapping them in conditional calls?

**Comment by mglaman** [comment]
Given `\PhpParser\NodeTraverser::ensureReplacementReasonable` I am not sure if this is possible.

**Comment by samsonasik** [comment]
Could you show the expected diff, before vs after?

**Comment by mglaman** [comment]
It would be something like the following. But I don't think it is possible with a Rule itself and requires a PostRector.

Existing:
```php
$password = user_password();
```

Replacement:
```php
    if (version_compare(\Drupal::VERSION, '9.1.0', '>=')) {
        $password = \Drupal::service('password_generator')->generate();
    }
    else {
        $password = user_password();
    }
```

The idea is to provide a replacement with backwards compatibility.

**Comment by samsonasik** [comment, marked as answer]
You need to replace whole `Expresssion` for that:

**First, define the node types as `Expression` stmt:**

```php
    public function getNodeTypes(): array
    {
        return [\PhpParser\Node\Stmt\Expression::class];
    }
```

**Next, verify if the node expr is match with `Assign`**

```php
    /**
     * @param \PhpParser\Node\Stmt\Expression $node
     */
    public function refactor(Node $node): ?Node
    {
        if (! $node->expr instanceof \PhpParser\Node\Expr\Assign) {
            return null;
        }

        if (! $node->expr->expr instanceof \PhpParser\Node\Expr\FuncCall) {
            return null;
        }

        if (! $this->nodeNameResolver->isName($node->expr->expr, 'user_password')) {
            return null;
        }

         // here replace with new If_
         $if = \PhpParser\Node\Stmt\If_(
               $this->nodeFactory->createFuncCall('version_compare', [
                     $this->nodeFactory->createClassConstFetch('Drupal', 'VERSION'), 
                     new \PhpParser\Node\Scalar\String_('9.1.0'),
                     new \PhpParser\Node\Scalar\String_('>='),
               ])
         );

         $if->stmts[0] = new \PhpParser\Node\Stmt\Expression(
                 // here new assign Expr...
         );
         $if->else = new \PhpParser\Node\Stmt\Else([
                   new \PhpParser\Node\Stmt\Expression(
                         $node->expr
                   ]),
         );

         return $if;
    }
```

Above is pseudo code, but you have the idea :)

**Reply by mglaman** [reply]
Thanks! I'll need to go back to the drawing board because I really just want to target replacing `user_password()` in all instances, but wrap as much as we can in backward compatibility.

**Reply by samsonasik** [reply]
One more think, you need to check `parent` node attribute already inside `Else_` to avoid infinite apply, as it will lookup another same `Expression` outside `Else_`:

```php
$else = $this->betterNodeFinder->findParentType($node, \PhpParser\Node\Stmt\Else::class);
if ($else instanceof \PhpParser\Node\Stmt\Else) {
     return null;
}
```
